source/fundamentals/data-formats/codecs.txt

```diff
@@ -11,38 +11,337 @@ Codecs
+instantiate the Codecs on which the ``Monolight`` depends including the
+``PowerStatusCodec`` from our prior example:
+
+.. literalinclude:: /includes/fundamentals/code-snippets/MonolightCodecProvider.java
+   :start-after: start class
+   :end-before: end class
+   :language: java
+   :dedent:
+
+To see a runnable example that demonstrates read and write operations using
+these Codecs, see the :ref:`Custom Codec Example <codecs-custom-example>`
+section of this guide.
+
+When working with POJOs, consider using the  ``PojoCodecProvider`` to
+minimize duplicate code to convert commonly-used data types and customize
+the behavior. See our
+:doc:`POJO Customization guide </fundamentals/data-formats/pojo-customization>`
+for more information.
+
+For more information about the classes and interfaces in this section, see the
+:java-docs:`CodecProvider API documentation <apidocs/bson/org/bson/codecs/configuration/CodecProvider.html>`.
 
 BsonTypeClassMap
 ~~~~~~~~~~~~~~~~
 
-- Description
-- Example of how to use to specify Bson mapping
+The ``BsonTypeClassMap`` class contains a recommended mapping between BSON
+and Java types. You can use this class to help you manage which Java types
+to decode your BSON types to in your custom ``Codec`` or ``CodecProvider``.
+You can add or modify the mapping by passing a ``Map`` containing new or
+replacement or entries. The ``Document`` class uses the recommended mapping.
+
+The following code snippet shows how you can retrieve the Java class type
+that corresponds to the BSON type in the default ``BsonTypeClassMap``
+instance:
+
+.. code-block:: java
+   :copyable: true
+
+   BsonTypeClassMap bsonTypeClassMap = new BsonTypeClassMap();
+   Class<?> clazz = bsonTypeClassMap.get(BsonType.ARRAY);
+   System.out.println("Java type: " + clazz.getName());
+
+This code outputs the following:
+
+.. code-block:: none
+   :copyable: false
+
+   Java type: java.util.List
+
+You can modify these mappings in your instance by specifying them in the
+``BsonTypeClassMap`` constructor. The following code snippet shows how
+you can replace the mapping for ``ARRAY`` in your ``BsonTypeClassMap``
+instance with the ``Set`` class:
+
+.. code-block:: java
+   :copyable: true
+
+   Map<BsonType, Class<?>> replacements = new HashMap<BsonType, Class<?>>();
+   replacements.put(BsonType.ARRAY, Set.class);
+   BsonTypeClassMap bsonTypeClassMap = new BsonTypeClassMap(replacements);
+
+   Class<?> clazz = bsonTypeClassMap.get(BsonType.ARRAY);
+   System.out.println("Java type: " + clazz.getName());
+
+This code outputs the following:
+
+.. code-block:: none
+   :copyable: false
+
+   Java type: java.util.Set
+
+For a complete list of the default mappings, refer to the
+:java-docs:`BsonTypeClassMap API documentation <apidocs/bson/org/bson/codecs/BsonTypeClassMap.html>`.
+
+For an example of how the ``Document`` class uses ``BsonTypeClassMap``, see
+the driver source code for the following classes:
```

Comment:
Should these source code links link to a line number? It is difficult to tell where in the source code you are referring to.

Reply:
I think it depends on factors including intent, feasibility, maintainability, and audience.

In this case:

- The intent is not to walk the user through the source code, but rather to mention they can trace the code themselves independently for further reading/learning. We've already explained one example of its usage above.

- For feasibility, in this case it's multiple lines of code, so linking to one line could be deceptive more than helpful, making the user believe that the reference might only be in one place.

- For maintainability, we need to be extra careful about linking to line numbers in source code since those can change in point releases without us knowing.

- As for audience, this paragraph is explicitly mentioning driver source code to allow users to determine whether this is the type of learning they want to pursue. 

Do you disagree with any of these points? And can you think of a better solution that addresses your request?

Reply:
From a technical perspective, I think we can circumvent the maintainability and feasibility concerns (points 2 and 3) by linking to multiple lines of a file at a specific commit hash. Example: https://github.com/mongodb/mongo-java-driver/blob/bd522224a66418c89595601263265da510843206/bson/src/main/org/bson/codecs/DocumentCodecProvider.java#L40-L63.

In regards to point 1, I think readers will have a more productive time tracing the source code with a definite starting point. Often for me the hardest thing about reading source code is gaining my first foothold. Having an explicit place to start/end has in the past been extremely helpful for me in beginning the process of studying an unfamiliar project.

In regards to point 4, I think there might have been a miscommunication due to me incorrectly selecting the lines I wished to comment on. The only things I think could benefit from a change are lines 268 and 269 such that the `href`s link to line numbers of a specific commit hash:

`- :github:`DocumentCodecProvider <mongodb/mongo-java-driver/blob/master/bson/src/main/org/bson/codecs/DocumentCodecProvider.java>`
- :github:`DocumentCodec <mongodb/mongo-java-driver/blob/master/bson/src/main/org/bson/codecs/DocumentCodec.java>``

Reply:
> In regards to point 4, I think there might have been a miscommunication due to me incorrectly selecting the lines I wished to comment on. The only things I think could benefit from a change are lines 268 and 269 such that the hrefs link to line numbers of a specific commit hash. These are the two lines I think could be changed:

I think those are the same links I'm referring to as they're the only ones that link to the driver repository. In your recommended lines to highlight, I'm not sure whether those are sufficient since the variable that contains the reference is used in more methods than the highlighted ones. I'm not sure a reader can understand the intent without explanation either.

Unfortunately, linking to a commit hash doesn't make it more maintainable than linking to a branch.  Since we're presenting documentation for specific versions, we would need to ensure we're showing them the "current implementation for this version" rather than "an implementation in a version, but not necessarily this version". Since a referenced commit is not guaranteed to be the same across PATCH versions, we can't guarantee those line numbers are valid in their MINOR version (for definitions of MINOR and PATCH, see semver.org).

Current issues that we've identified in this conversation:
- the user might not find the links to source code helpful without more guidance
- as document creators, it is difficult to guide some users through all the details of the source code (especially with just line numbers).
- as document creators, it's difficult to come up with a way to make documentation that guides a user through specific lines of source code maintainable

Initially, I thought it could be helpful, but I think from hearing your feedback, maybe I should do one or more of the following:

1) omit the link to the source code
2) replace it with a contrived example of how it might be able to used in the decode() method of a Codec
3) omit the section

I'll see if someone on the Java team wants to weigh in on what their thoughts are in an external review.

Reply:
After hearing your response, I agree that linking to a specific commit hash is not a good idea. I like the problems you have outlined. How to effectively utilize source code as a teaching tool could be a worthwhile conversation with DBX engineering teams. Maybe certain sections of the source are heavily commented and designed to be read as educational material? Not sure how to solve this. 

Resolving this comment as I am no longer requesting any changes.